### PR TITLE
Templating for ClusterCIDR and IPs in addons

### LINKED ISSUE #1167: Add templating for pod&service cidr for Addons
Take network specs from the Cluster object and use it within the Addon-templates.
Make sure:
- Pod cidr gets correctly set in Flannel/Canal
- Cluster IP of DNS is within the service range

### PR DESCRIPTION
**Pull request by glower**
**What this PR does / why we need it**:

* Check if cluster IP of DNS is within the service range

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1167

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: api/pkg/controller/addon/controller_test.go [hunk 1]
```diff
@@ -121,20 +137,66 @@ func setupTestCluster() *kubermaticv1.Cluster {
...
+		addonDir: addonDir,
+	}
+	manifests, err := controller.getAddonManifests(addon, cluster)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(manifests) != 1 {
+		t.Fatalf("invalid number of manifests returned. Expected 1, Got %d", len(manifests))
+	}
+	expectedIP := "10.10.10.1"
```
Comment (mrIncompetent) [inline api/pkg/controller/addon/controller_test.go]: The first ip of the CIDR is reserved for the Apiserver. As mentioned please use the 10.

Comment (glower) [reply api/pkg/controller/addon/controller_test.go]: done

File: api/pkg/controller/addon/controller.go [hunk 2]
```diff
@@ -297,6 +298,8 @@ type templateData struct {
 	Cluster           *kubermaticv1.Cluster
 	Variables         map[string]interface{}
 	OverwriteRegistry string
+	KubeDNSclusterIP  string
```
Comment (mrIncompetent) [inline api/pkg/controller/addon/controller.go]: Rename to `DNSClusterIP`

File: api/pkg/controller/addon/controller.go [hunk 3]
```diff
@@ -297,6 +298,8 @@ type templateData struct {
...
+	ClusterCIDR       string
```
Comment (mrIncompetent) [inline api/pkg/controller/addon/controller.go]: Why introducing a dedicated field when its already accessable via the `.Cluster` ?

File: addons/dns/kube-dns.yaml [hunk 4]
```diff
@@ -30,7 +30,7 @@ metadata:
 spec:
   selector:
     k8s-app: kube-dns
-  clusterIP: 10.10.10.10
+  clusterIP: {{.KubeDNSclusterIP}}
```
Comment (mrIncompetent) [inline addons/dns/kube-dns.yaml]: Rename to `DNSClusterIP`

**Review by mrIncompetent — CHANGES_REQUESTED** [review]

File: api/pkg/controller/addon/controller.go [hunk 5]
```diff
@@ -308,11 +311,18 @@ func (c *Controller) getAddonManifests(addon *kubermaticv1.Addon, cluster *kuber
...
+		return nil, err
+	}
+
 	data := &templateData{
 		Variables:         make(map[string]interface{}),
 		Cluster:           cluster,
 		Addon:             addon,
 		OverwriteRegistry: c.registryURI,
+		KubeDNSclusterIP:  kubeDNSclusterIP,
+		ClusterCIDR:       cluster.Spec.ClusterNetwork.Pods.CIDRBlocks[0],
```
Comment (kron4eg) [inline api/pkg/controller/addon/controller.go]: That `CIDRBlocks[0]` is a potential panic (when the slice is empty) waiting to happen.

Comment (glower) [reply api/pkg/controller/addon/controller.go]: Hmm, OK I can check it, but if Pods.CIDRBlocks is empty, the cluster will probably not run anyway

Comment (glower) [reply api/pkg/controller/addon/controller.go]: I will move this part now to the template anyway

File: api/pkg/controller/addon/controller.go (hunk 3 above)
Comment (glower) [reply api/pkg/controller/addon/controller.go]: You right

File: api/pkg/controller/addon/controller.go (hunk 2 above)
Comment (glower) [reply api/pkg/controller/addon/controller.go]: fixed

File: addons/dns/kube-dns.yaml (hunk 4 above)
Comment (glower) [reply addons/dns/kube-dns.yaml]: fixed

**Review by mrIncompetent — APPROVED** [review]

**Pull request merged**